chia/consensus/blockchain_mmr.py

```diff
@@ -0,0 +1,249 @@
+from __future__ import annotations
+
+import logging
+
+from chia_rs import BlockRecord
+from chia_rs.sized_bytes import bytes32
+from chia_rs.sized_ints import uint32
+
+from chia.consensus.blockchain_interface import BlockRecordsProtocol
+from chia.consensus.mmr import MerkleMountainRange
+
+log = logging.getLogger(__name__)
+
+
+class BlockchainMMRManager:
+    """
+    Manages MMR state for blockchain operations.
+    Includes checkpointing for efficient rollback during reorgs.
+    """
+
+    _mmr: MerkleMountainRange
+
+    _checkpoints: dict[int, MerkleMountainRange]  # height -> MMR snapshot
+    _checkpoint_interval: int
+    _max_checkpoints: int
+
+    def __init__(
+        self, mmr: BlockchainMMRManager | None = None, checkpoint_interval: int = 1000, max_checkpoints: int = 10
+    ) -> None:
+        # Current MMR state
+        if mmr is not None:
+            self._mmr = mmr._mmr.copy()
+            self._last_header_hash: bytes32 | None = mmr._last_header_hash
+            self._last_height: uint32 | None = mmr._last_height
+            self._checkpoints = {h: mmr_snap.copy() for h, mmr_snap in mmr._checkpoints.items()}
+            self._checkpoint_interval = mmr._checkpoint_interval
+            self._max_checkpoints = mmr._max_checkpoints
+        else:
+            self._mmr: MerkleMountainRange = MerkleMountainRange()
+            self._last_header_hash = None
+            self._last_height = None
+            self._checkpoints: dict[int, MerkleMountainRange] = {}
+            self._checkpoint_interval = checkpoint_interval
+            self._max_checkpoints = max_checkpoints
+
+    def copy(self) -> BlockchainMMRManager:
+        return BlockchainMMRManager(self)
+
+    def add_block_to_mmr(self, header_hash: bytes32, prev_hash: bytes32, height: uint32) -> None:
+        """
+        Add a block to the MMR in sequential order.
+        This should be called for blocks in height order to maintain MMR integrity.
+        """
+
+        # Only add blocks that are the next expected height
+        if self._last_header_hash is not None and (prev_hash != self._last_header_hash):
+            # Skip blocks that are out of order or duplicate
+            log.warning(
+                f"Skipping block height {height}, prev_hash mismatch "
+                f"(expected {self._last_header_hash.hex()[:16]}, got {prev_hash.hex()[:16]})"
```

Comment:
this this expected to happen? it looks a bit too serious to just log an move on.